File: packages/gbnf-experiment/data/20260909T193332Z_f85fdfbb/ported_implementation/utils/errors/grammar_parse_error.py
from __future__ import annotations

from .build_error_position import build_error_position


def GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason: str) -> str:
    return f"Failed to parse grammar: {reason}"


class GrammarParseError(Exception):
    """Raised when a grammar cannot be parsed."""

    def __init__(self, grammar: str, pos: int, reason: str):
        message = "\n".join([
            GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
            "",
            *build_error_position(grammar, pos),
        ])
        super().__init__(message)
        self.message = message
        self.name = "GrammarParseError"
        self.grammar = grammar
        self.reason = reason
        self.pos = pos

    def __str__(self) -> str:
        return self.message

File: proposed-projects/round-trip-experiment/reverse/20260909T012246Z_fbe56704/20260910T051241Z_fa3c6dd0/ported_implementation/src/utils/errors/grammar-parse-error.ts
import { buildErrorPosition } from './build-error-position.ts';

export const GRAMMAR_PARSER_ERROR_HEADER_MESSAGE = (reason: string): string =>
  `Failed to parse grammar: ${reason}`;

export class GrammarParseError extends Error {
  grammar: string;
  reason: string;
  pos: number;

  constructor(grammar: string, pos: number, reason: string) {
    super(
      [
        GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
        '',
        ...buildErrorPosition(grammar, pos),
      ].join('\n')
    );
    this.name = 'GrammarParseError';
    this.grammar = grammar;
    this.reason = reason;
    this.pos = pos;
  }
}
